hazelcast/src/main/java/com/hazelcast/map/impl/query/QueryEngineImpl.java

```diff
@@ -172,47 +187,61 @@ private Result populateResult(Query query, PartitionIdSet partitionIds) {
-                if (!finishedPartitionIds.containsAll(queriedPartitionIds)) {
-                    // do not take into account results that contain partition IDs already removed from partitionIds
-                    // collection as this means that we will count results from a single partition twice
-                    // see also https://github.com/hazelcast/hazelcast/issues/6471
-                    continue;
-                }
-                finishedPartitionIds.removeAll(queriedPartitionIds);
-                result.combine(queryResult);
-            }
-        }
+    private CompletableFuture<Result> addResultsOfPredicate(List<CompletableFuture<Result>> futures, Result result,
+                                                            PartitionIdSet finishedPartitionIds, boolean rethrowAll) {
+
+        List<CompletableFuture<Result>> futuresWithHandle = futures.stream()
+                .map(future -> {
+                            future = future.handle(
+                                    (queryResult, t) -> {
+                                        if (t != null) {
+                                            if (t.getCause() instanceof QueryResultSizeExceededException
+                                                    || rethrowAll) {
+                                                throw rethrow(t);
+                                            }
+                                            logger.fine("Could not get query results", t);
+                                        }
+                                        return queryResult;
+                                    }
+                            );
+                            return future;
+                        }
+                ).collect(Collectors.toList());
+
+        return CompletableFuture.allOf(futuresWithHandle.toArray(new CompletableFuture[futuresWithHandle.size()]))
+                .thenApply(v -> futuresWithHandle.stream().map(future -> future.join()).collect(Collectors.toList()))
+                .thenApply(listOfQueryResults -> {
+                    for (Result queryResult : listOfQueryResults) {
+                        if (queryResult == null) {
+                            continue;
+                        }
+                        PartitionIdSet queriedPartitionIds = queryResult.getPartitionIds();
+                        if (queriedPartitionIds != null) {
+                            if (!finishedPartitionIds.containsAll(queriedPartitionIds)) {
+                                // do not take into account results that contain partition IDs already removed from partitionIds
+                                // collection as this means that we will count results from a single partition twice
+                                // see also https://github.com/hazelcast/hazelcast/issues/6471
+                                continue;
+                            }
+                            finishedPartitionIds.removeAll(queriedPartitionIds);
```

Comment:
It might be a good time to give a new name to `finishedPartitionIds`, to my understanding the usage is opposite to the naming, it's like "unfinished" or "left" partitions.